cmd/eudico-stats/listen.go

```diff
@@ -0,0 +1,379 @@
+package main
+
+import (
+	"context"
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/api/v1api"
+	"github.com/filecoin-project/lotus/blockstore"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical/actors/sca"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical/actors/subnet"
+	"github.com/filecoin-project/lotus/chain/events"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/cmd/eudico-stats/observer"
+	cbor "github.com/ipfs/go-ipld-cbor"
+)
+
+const Timeout = 76587687658765876
+const Confidence = 0
+
+type SubnetStat struct {
+	// The nextEpoch to resync stats with the chain
+	nextEpoch abi.ChainEpoch
+	// The list of subnets currently managed by the current subnet
+	subnets map[address.SubnetID]bool
+	// The node to be stored in db
+	node SubnetNode
+}
+
+func emptySubnetStat(id address.SubnetID) SubnetStat {
+	return SubnetStat{
+		nextEpoch: 0,
+		subnets:   make(map[address.SubnetID]bool),
+		node: idOnlySubnetNode(id),
+	}
+}
+
+// ShouldReSync checks if it is needed to resync subnet stats with the current chain data
+func (s *SubnetStat) ShouldReSync(curH abi.ChainEpoch) bool {
+	return s.nextEpoch < curH
+}
+
+type EudicoStatsListener struct {
+	Events *events.Events
+
+	// Private fields
+	api         v1api.FullNode
+	observer    observer.Observer
+	subnetStats map[address.SubnetID]SubnetStat
+}
+
+func NewEudicoStats(ctx context.Context, api v1api.FullNode, observer observer.Observer) (EudicoStatsListener, error) {
+	eventListen, err := events.NewEvents(ctx, api)
+	if err != nil {
+		return EudicoStatsListener{}, err
+	}
+
+	listener := EudicoStatsListener{
+		Events:      eventListen,
+		api:         api,
+		subnetStats: make(map[address.SubnetID]SubnetStat),
+		observer:    observer,
+	}
+
+	log.Infow("Initialized eudico stats")
+
+	return listener, nil
+}
+
+// Listen TODO: placeholder for future public invocation differences compared to `listen`
+func (e *EudicoStatsListener) Listen(ctx context.Context, id address.SubnetID) error {
+	return e.listen(ctx, id)
+}
+
+func (e *EudicoStatsListener) listen(
+	ctx context.Context,
+	id address.SubnetID,
+) error {
+	// TODO: ideally there is no need to add read write lock. Keep in view.
```

Comment:
You may need a lock here, what if the listener accesses the map because you start syncing with a new subnet while you are reading from it to collect some metrics? I'd suggest running with `race` to see if this is the case.